src/main/scala/edu/ie3/simona/model/grid/GridModel.scala

```diff
@@ -592,6 +634,133 @@ case object GridModel {
+      vMin: Double
+  ): ControlGroupModel = {
+    val nodeUuids =
+      determineNodeUuids(measurementUnitInput, measurementConfigs)
+    buildTransformerControlModels(nodeUuids, vMax, vMin)
+  }
+
+  /** Determine the uuids of the nodes to control
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @return
+    *   A set of relevant nodal uuids
+    */
+  private def determineNodeUuids(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String]
+  ): Set[UUID] = Set.from(
+    measurementUnitInput.asScala
+      .filter(input =>
+        measurementConfigs.contains(input.getUuid.toString) && input.getVMag
+      )
+      .map(_.getNode.getUuid)
+  )
+
+  /** Build a single control group model. Currently, only limit violation
+    * prevention logic is captured: The nodal regulation need is equal to the
+    * voltage change needed to comply with the given thresholds
+    *
+    * @param nodeUuids
+    *   Collection of all relevant node uuids
+    * @param vMax
+    *   Upper permissible voltage magnitude
+    * @param vMin
+    *   Lower permissible voltage magnitude
+    * @return
+    *   A [[ControlGroupModel]]
+    */
+  private def buildTransformerControlModels(
+      nodeUuids: Set[UUID],
+      vMax: Double,
+      vMin: Double
+  ): ControlGroupModel = {
+    /* Determine the voltage regulation criterion for each of the available nodes */
+    val nodeUuidToRegulationCriterion = nodeUuids.map { uuid =>
+      uuid -> { (complexVoltage: Complex) =>
+        {
+          val vMag = complexVoltage.abs
+          if (vMag > vMax)
+            Some(vMax - vMag)
+          else if (vMag < vMin)
+            Some(vMin - vMag)
+          else
+            None
+        }.map(Quantities.getQuantity(_, PowerSystemUnits.PU))
+      }
+    }.toMap
+
+    val harmonizationFunction =
+      (regulationRequests: Array[ComparableQuantity[Dimensionless]]) => {
+        val negativeRequests = regulationRequests.filter(
+          _.isLessThan(Quantities.getQuantity(0d, PowerSystemUnits.PU))
+        )
+        val positiveRequests = regulationRequests.filter(
+          _.isGreaterThan(Quantities.getQuantity(0d, PowerSystemUnits.PU))
+        )
+
+        if (negativeRequests.nonEmpty && positiveRequests.nonEmpty) {
+          /* There are requests for higher and lower voltages at the same time => do nothing! */
+          None
+        } else if (negativeRequests.nonEmpty) {
+          /* There are only requests for lower voltages => decide for the lowest required voltage */
+          negativeRequests.minOption
+        } else {
+          /* There are only requests for higher voltages => decide for the highest required voltage */
+          positiveRequests.maxOption
+        }
+      }
+
```

Comment:
**Double-check this section:** To my understanding, if vMax is set to 12 p.u. and there is a higher voltage request, there is a possibility for the voltage to be set at 12 p.u. (hypothetically speaking). Or did I get this wrong?

Reply:
@sensarmad: Not sure if I got this right. In case vMax is (hypothetically) set to 12 p.u. then only nodes with absolute voltage over 12 p.u. would activate the regulation. With new the boundaries vMax is now limited hard till 1.2 (and vMin till 0.8). So if vMax is now set to 1.1 p.u. and the actual voltage is for one node at 1.15 and for another at 1.12, then the regulation will identify just positive deviations (of 0.05 and 0.03 to vMax and will choose the maximum of these (here 0.05) as the regulation need.